tools/cbmc/proofs/DNS/DNSgetHostByName/DNSgetHostByName_harness.c

```diff
@@ -0,0 +1,47 @@
+/* FreeRTOS includes. */
+#include "FreeRTOS.h"
+#include "queue.h"
+
+/* FreeRTOS+TCP includes. */
+#include "FreeRTOS_IP.h"
+#include "FreeRTOS_DNS.h"
+#include "FreeRTOS_IP_Private.h"
+#include "FreeRTOS_Sockets.h"
+
+/* This assumes the length of pcHostName is bounded by MAX_HOSTNAME_LEN and the size of UDPPayloadBuffer is bounded by 
+MAX_REQ_SIZE. */ 
+
+void *safeMalloc(size_t xWantedSize) {
+	if(xWantedSize == 0) {
+		return NULL;
+	}
+	uint8_t byte;
+	return byte ? malloc(xWantedSize) : NULL;
+}
+
+/* Abstraction of FreeRTOS_GetUDPPayloadBuffer. This should be checked later. For now we are allocating a fixed sized memory of size MAX_REQ_SIZE. */
+void * FreeRTOS_GetUDPPayloadBuffer(size_t xRequestedSizeBytes, TickType_t xBlockTimeTicks ) {
+	void *pvReturn = safeMalloc(MAX_REQ_SIZE); 
+	return pvReturn;
+}
+
+/* Abstraction of FreeRTOS_socket. This abstraction allocates a memory of size Socket_t. */
+Socket_t FreeRTOS_socket( BaseType_t xDomain, BaseType_t xType, BaseType_t xProtocol ) {
+	Socket_t xSocket = safeMalloc(sizeof(Socket_t)); /* Replaced malloc by safeMalloc */
+	return xSocket;
+}
+
+/* This function only uses the return value of prvParseDNSReply. Hence it returns an unconstrained uint32 value */
+uint32_t prvParseDNSReply( uint8_t *pucUDPPayloadBuffer, size_t xBufferLength, TickType_t xIdentifier) { }
+
+void harness() {
+	size_t len;
+	__CPROVER_assume(len >= 0 && len <= MAX_HOSTNAME_LEN);
+	char *pcHostName = safeMalloc(len); /* Replaced malloc by safeMalloc */
+	if (len && pcHostName) {
+		pcHostName[len-1] = NULL;
```

Comment:
I've seen this in a couple of these now.  Is this accurately expressing an invariant of the original source, or is this introducing an assumption?  Seems to me it would be desirable to catch the case when the sentinel character is not present.

Reply:
This is required as there are `strlen`, `strcpy` functions in the code which require `NULL` termination. In this harness we are assuming any `NULL` terminated  `hostname` of length up to `MAX_HOSTNAME_LEN`. If we want to catch the case where the `hostname` is not `NULL` terminated, we need to assert it in `FreeRTOS_DNS.c`.

Reply:
I see.  So, the proof is only valid when this precondition is met?

Reply:
Yes.